src/passes/OptimizeInstructions.cpp

```diff
@@ -910,6 +916,28 @@ struct OptimizeInstructions
         swap();
       }
     };
+    auto hasConstantOnRight = [](Expression* expr) {
+      // Check if expression has constant on the right.
+      // Values such as -1 or 0 always prefer on a right since commonly
+      // found in pattern matching
```

Comment:
I apologize for repeating myself, but I still disagree on this special-casing of 0 and -1. Let me try to explain why I think this matters, and why I am not just being pedantic.

I need to have a clear mental model of what canonicalization means. When I look at two expressions, it should be obvious what the canonical form is. "0 and -1 are commonly found in pattern matching" doesn't really explain enough to me.

First, why is 1 not in that list? If any constant is commonly found, it should be 1, shouldn't it?

Second, why even pick 0, -1 (and maybe 1), and not just do this for all constants? **If there is no downside to doing it for all constants**, then we should do that, as the mental model will be simpler, and the canonicalization will handle more cases - which is always better. (If there is a downside, please specify it here!)

Reply:
Sorry. May be I should clarify better. `-1` and `0` on the right only when lhs values also constant but not equal `-1` or `0`. So in any case we get any constant on a right side for sub-expressions, but for if we got `x + (1234 ^ -1)` we make sure it will not be swap as `x + (-1 ^ 1234)`. Ofc we could just ignore this and it's not big deal for next passes which just fold this const expressions, but in my experiments if we ignore this cases we got a lot of unnecessary swaps. And `C ^ -1` is pretty common because it actually bitwise not `~C` and could interpret as unary operation.

Reply:
Better example. Let's say we have the following expression:
```
(x + 1) & ((x + 1) ^ -1)
```
which we could simplify to:
```
Y & (Y ^ -1)
```
where `Y = x + 1`
and after simplification it becomes just zero.

In this canonical form we just need write one pattern match and don't need check this case `(Y ^ -1) & Y`

But if we don't care about and accidental swap subexpression with `-1` to this:
```
((x + 1) ^ -1) ^ (x + 1)
```
We got `(Y ^ -1) & Y` and should do double check

Reply:
btw LLVM it seems do the same thing: https://godbolt.org/z/95F4kZ

Reply:
LLVM do much complex canonization:
https://github.com/llvm/llvm-project/blob/f39e12a06b6018db195848ca1f7bd01bf0240fac/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L305

And use [complexity](https://github.com/llvm/llvm-project/blob/master/llvm/lib/Transforms/InstCombine/InstCombineInternal.h#L84) for determine priority.`Neg/Not` expressions have second-to-last priority which higher than constant expressions. I guess we could do something similar.

Reply:
Why not canonicalize using a more general rule, like if there is a constant on both sides, prefer the lower constant on the left?

Again, the goal should be, I think, to make a simple rule that handles as many cases. By simple I mean, "does not raise obvious questions like why this or that constant". Sorting lower constants to the left (or right, doesn't matter as long as there is a rule on way or the other) doesn't raise obvious questions, as it just picks a side for them, nothing is special-cased.

Reply:
Or, lower constant in absolute value, or some other sorting may be even better - just without special-casing of particular constants, unless there is a real reason for it we can document.

I don't think "we have some rules that can optimize -1, so special-case it" is convincing enough - because the obvious question raised there is, why not generalize the rule even more. That seems to have no downside, and the more general the better.

Reply:
Currently I'm not specialize constants. Just detect bitwise negative (~x) operation which in wasm represent as `x ^ -1` but include in many optimization patterns. Also detect `-x` which is `0 - x`. So current strategy little bit different and more aligned with LLVM approach.

Reply:
In this comment I described new general approach:
https://github.com/WebAssembly/binaryen/pull/2894/files#diff-54e02686df3480c98e3f90a24705c00bR914

Reply:
so my first assumption was we need care about `-1` and `0` constant but without any context. But currently I bind this constants to concrete operation and now it has semantics. Unfortunately wasm hasn't special `neg` and `not` operations so we need detect such binary patterns and actually `x ^ -1` and `0 - x` could interpret as pseudo-unary operations. Is it make sense?

Reply:
LLVM sort sub-expressions of top commutative expression by complexity (from most complex to most simplest expression which is constant). So I try to do something similar.

May be @tlively who know LLVM better than me suggest something more clever?
I am happy to hear any suggestion that will simplify pattern matching.

Reply:
Basically I don't try replicate exact LLVM's behaviour which much more complex and perhaps more expensive, but inherit most useful for parts which already pretty welcomed and could be helpful for this [PR](https://github.com/WebAssembly/binaryen/pull/2884) for example.

Alright I could simplify `computeExpressionCost` and don't care about `VExpr` currently so it slightly simplify logic

Reply:
With recent commit it currently looks like this:
```
(BExpr|UExpr) > BCExpr > (XExpr|NExpr|ZExpr) > UCExpr > CExpr
       -           -              2              1        1
                         |----------------------------------|
                          handled by `computeExpressionCost`

```

Reply:
@MaxGraey 

I think I see what you're saying better now. So you want to find patterns with `0, -1` and for that you want them to be on the left (say). A more general rule that sorts all the constants might not do that, for example if it sorts by the size of the constant, then we'd have `-2, 0` but `0, 5` (so `0` is not always on the same side).

> The main task is to canonize in such a way that it would be easier to pattern matching, isn't it? And only the second goal is codegen determinism.

I wouldn't say "determinism" here since "nondeterministic build" means something else (run it more than once, get different output). But yes, that's a good point to distingish two different goals here. One is pattern matching, and the other is to avoid redundancy - to always, as much as possible, have a single way to represent things, without symmetrical permutations that just add complexity.

Usually avoiding redundancy is enough to allow easy pattern matching. But if the patterns specifically have `0, -1` then it isn't.

I'm convinced that we may want to special-case those. However, I think we'll want to do more eventually, like special-case 1 and 2 and others. So I hope we can find a more general solution here.

One option might be to sort by `0, -1, 1, -2, 2, -3, 3` etc. - a simple deterministic order that puts the constants you want at the top. Another option might be to give up on canonicalization for this, if it requires this complexity, and to instead look for these `neg / not` patterns on both sides of a binary. I'm not sure which I prefer.

Reply:
Still not sure is it make sense do special sorting. It could quite expensive and made logic much complex. Even LLVM don't care abut this: https://godbolt.org/z/DpgkgJ

Reply:
Yes, maybe this is too complex to sort.

So maybe we should give up on canonicalization for this. We can return to your PR that motivated this, and use a helper function to check for your patterns on both the left and the right.

Reply:
Yeah, what if we put this PR on hold until all the patterns are clear during next PR developing? I would like to do the following PRs in pending list. This PR has caused too much discussion)

Reply:
Also you are free to make your own PR which improve canonization which better in your opinion. I'm totally fine with this case as well

Reply:
Sounds good, yes, let's put this PR on hold (feel free to close it or mark as draft) and get back to the next in your main list of PRs (please ping me there)